com.unity.shadergraph/Editor/GraphUI/GraphElements/RegistryNode.cs

```diff
@@ -0,0 +1,30 @@
+using UnityEditor.GraphToolsFoundation.Overdrive;
+using UnityEngine.UIElements;
+
+namespace UnityEditor.ShaderGraph.GraphUI
+{
+    /// <summary>
+    /// A RegistryNode is the view for a RegistryNodeModel. It gets a description of its UI from the ShaderGraph
+    /// Registry.
+    /// </summary>
+    public class RegistryNode : CollapsibleInOutNode
+    {
+        protected override void BuildPartList()
+        {
+            base.BuildPartList();
+
+            if (Model is not RegistryNodeModel registryNodeModel) return;
+
+            // TODO: Build part list from node definition
```

Comment:
Wondering what this might look like more formally...

if  `NodeModel` exposes the actual NodeDefinition from registry (as it contains the registry key for lookup)

```
var nodeDefinition = Model.GetNodeDefinition();
foreach(var nodeDefinitionPart : nodeDefinition.Parts)
{
      if(nodeDefinitionPart.IsInput)
            AddDataInputPort(nodeDefinitionPart.name, nodeDefinitionPart.TypeHandle, .....)
      else
            AddDataOutputPort(nodeDefinitionPart.name, nodeDefinitionPart.TypeHandle, .....)
            
}
```

Will clarify with Esme what to expect in terms of the Node Topology and Type Information we can retrieve from the Registry

Reply:
Also, that has me thinking about where we will define whether a node has a preview or not attached to it, probably makes sense for it to be part of the `NodeDefinition.Parts` ? 

Then we can just react to it and add a 'NodePreviewPart` to the node part list in response

Curious on your thoughts here

Reply:
Starting with the ports specifically: those will be handled in `RegistryNodeModel.OnDefineNode` with a loop similar to your sample. Based on what I understand, the registry will have a deterministic function that maps (node type, node connections) -> topology for us to use. GTF automatically makes port parts for us, so that's already accounted for.

Now, other node parts are a bit more interesting. The registry could give us a list of parts (i.e. swizzle node -> [mask text field, preview]), but because the registry isn't responsible for anything GTF, we'd have to figure out how to map its concepts into visual elements. This sounds like something we'd want to discuss further as a whole -- I'm not actually sure where the responsibility should be.